python/ray/tune/trial_runner.py

```diff
@@ -480,25 +502,53 @@ def _process_trial(self, trial):
             self._checkpoint_trial_if_needed(
                 trial, force=result.get(SHOULD_CHECKPOINT, False))
 
-            if decision == TrialScheduler.CONTINUE:
-                self.trial_executor.continue_training(trial)
-            elif decision == TrialScheduler.PAUSE:
-                self.trial_executor.pause_trial(trial)
-            elif decision == TrialScheduler.STOP:
-                self.trial_executor.export_trial_if_needed(trial)
-                self.trial_executor.stop_trial(trial)
+            if trial.is_saving:
+                # Cache decision to execute on after the save is processed.
+                # This prevents changing the trial's state or kicking off
+                # another training step prematurely.
+                self._cached_trial_decisions[trial.trial_id] = decision
             else:
-                assert False, "Invalid scheduling decision: {}".format(
-                    decision)
+                self._execute_action(trial, decision)
         except Exception:
             logger.exception("Trial %s: Error processing event.", trial)
             self._process_trial_failure(trial, traceback.format_exc())
 
+    def _process_trial_save(self, trial):
+        """Processes a trial save.
+
+        Acts on the decision cached during the last `_process_trial` call.
+
+        Args:
+            trial (Trial): Trial being saved.
+        """
+        logger.debug("Trial %s: Processing trial save.", trial)
+        checkpoint_value = None
+
+        try:
+            checkpoint_value = self.trial_executor.fetch_result(trial)
+        except Exception:
```

Comment:
Should these exception catches be a bit more strict? Right now it seems like things can just silently fail. I could imagine that in many cases failing to checkpoint is worth just failing the whole trial.

Reply:
yeah the exception catching should definitely be more strict in general in trial_runner but I'd leave that to another PR. 

To your second point, if the save function fails due to user error etc the trial is failed explicitly. It only doesn’t if the error is due to something like the rsync failing but even then the error will be logged. are you proposing we should fail explicitly there as well (btw in this case an error is still logged)

Reply:
Yeah, it's not obvious that failing in case of rsync errors is actually the right thing to do in general, but I can definitely see cases where not having saved checkpoints is as bad (or even worse) than just failing the complete run.

Reply:
what what I understand, rsync failure behavior does not change given this PR - should we first merge this and work on a solution in a later PR?

I actually think the right thing might be to migrate all data movement through Ray on default and not use rsync for inter-node communication.

Reply:
Or actually more specifically, discuss this question in an issue and ping other core Ray folks to ask them if there is a thing we should generally support in Ray?

Reply:
yeah this PR doesnt change any error-handling behavior so lets discuss it separately. and yeah this won't be as much of an issue once we deprecate rsync support / finalize the durable API

Reply:
cc @hartikainen in case any objections here

Reply:
Oh yeah, sounds good! Sorry, did not mean to delay this PR with the original comment.